Remove WC Core shipping settings if Cart/Checkout blocks are in use

opr opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This is part of a series of PRs required to satisfy #7895.

In this PR we remove settings from `WooCommerce -> Settings -> Shipping -> Shipping Options`

Specifically the options disabled are:

- `Enable the shipping calculator on the cart page` if the Cart Block is present on the Cart page and,
- `Hide shipping costs until an address is entered` if the Checkout Block is present on the Checkout page.

The first setting is controlled by the Cart Block's settings, and the second option should always be `no` for the purposes of getting the correct local pickup rates from the store API, and will be moved to a block setting in a later PR.

I also added a utility function to see if the Cart/Checkout blocks are the default experience. This check could be made multiple places over the codebase, so when we spot it we can update it to use the utility function.

It also adds E2E tests to ensure this is working as expected.

### Screenshots

<!-- If your change has a visual component, add a screenshot here. A "before" screenshot would also be helpful. -->

| Before | After |
| ------ | ----- |
| <img width="612" alt="image" src="https://user-images.githubusercontent.com/5656702/223591261-a40e6741-43a2-458a-af17-250c1141ae94.png"> |  <img width="611" alt="image" src="https://user-images.githubusercontent.com/5656702/223591191-143578f1-cfc3-46ad-9b67-15f4f4484ca9.png"> |

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Go to `WooCommerce -> Settings -> Advanced` - set the Cart and Checkout pages to pages containing the **shortcode** cart/checkout experience. Save.
2. Go to `WooCommerce -> Settings -> Shipping -> Shipping Options` - observe the two options, `Hide shipping costs until an address is entered` and `Enable the shipping calculator on the cart page`.
3. Go to `WooCommerce -> Settings -> Advanced` - set the **Cart** page to one containing the **Cart Block**. Save.
4. Go to `WooCommerce -> Settings -> Shipping -> Shipping Options` - observe that where the two options were in step 2, only, `Hide shipping costs until an address is entered` remains. Ensure it displays correctly and looks OK.
5. Go to `WooCommerce -> Settings -> Advanced` - set the **Checkout** page to one containing the **Checkout Block**. Save.
6. Go to `WooCommerce -> Settings -> Shipping -> Shipping Options` - observe that where the two options were in step 2 nothing appears.
7. Go to `WooCommerce -> Settings -> Advanced` - set the **Cart** page to one containing the **Shortcode cart**. Save.
8. Go to `WooCommerce -> Settings -> Shipping -> Shipping Options` - observe that where the two options were in step 2 only `Enable the shipping calculator on the cart page` appears. Ensure it displays correctly.

Expected results

| | Shortcode Cart | Block Cart |
|-|----------------|------------|
| **Shortcode Checkout** | <img width="613" alt="image" src="https://user-images.githubusercontent.com/5656702/223882079-ac03e6ef-f1e9-4f7b-b173-65752866030e.png"> | <img width="582" alt="image" src="https://user-images.githubusercontent.com/5656702/223882155-b8dd6916-5213-4be9-9cae-b04e850a4d3f.png"> |
| **Block Checkout** | <img width="579" alt="image" src="https://user-images.githubusercontent.com/5656702/223882325-87a77a3c-5409-485d-ad5d-e422273f5d66.png"> | <img width="538" alt="image" src="https://user-images.githubusercontent.com/5656702/223882267-d0c27ae2-01e7-4691-9264-7101051ff48c.png"> |
 
* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

### Changelog

> Remove certain Shipping settings from WooCommerce -> Settings -> Shipping -> Shipping Options when using the Cart or Checkout blocks, these have been moved to setting on the blocks.

opr committed:
Add CartCheckoutUtils class

This class will store reusable methods relating to Cart/Checkout Blocks, i.e. whether they are used on the Cart/Checkout page.

opr committed:
Update ShippingController to use the new CartCheckoutUtils function

This will reduce code duplication when checking if the Cart/Checkout blocks are in use on the Cart/Checkout page.

opr committed:
Add filter to remove shipping settings when Cart/Checkout are default

opr committed:
Force shipping to be enabled if the Checkout block is in use.

opr committed:
Add setCartCheckoutPages function to update set the cart/checkout page

opr committed:
Add setCartCheckoutPages function to update set the cart/checkout page

opr committed:
Add tests to ensure core shipping settings update correctly

opr committed:
Add filter to override cost requires address option

opr committed:
Add shippingCostRequiresAddress option

opr committed:
Add isAddressComplete function

Borrowed from #8141

opr committed:
Revert "Add isAddressComplete function"

This reverts commit 9967dc0d4f10cf638859ae085e6f4cc2901dd299.

opr committed:
Check if the address is required before showing rates

opr committed:
Check if the address is required before showing rates

opr committed:
Show shipping rates in editor

opr committed:
Show shipping rates in editor

opr committed:
Add shippingCostRequiresAddress attribute to shipping methods block

opr committed:
Update frontend type to show shippingCostRequiresAddress is a prop

opr committed:
Add control to toggle shippingCostRequiresAddress option

opr committed:
Show address notice in the correct scenario

opr committed:
Show address notice in the correct scenario

opr committed:
Send shippingCostRequiresAddress to Block in front end context

opr committed:
Add e2e test for editor control

opr committed:
Add e2e tests for shipping options on the front end

opr committed:
Add e2e tests for shipping options on the front end

opr committed:
Ensure errorId is passed to StateInput

opr committed:
Add updateAttributeInSiblingBlock function

opr committed:
Add shippingCostRequiresAddress to shipping method block

opr committed:
Ensure attribute is updated in both blocks when editing

opr committed:
Add fullShippingAddressPushed action to wc/store/cart

opr committed:
Add fullShippingAddressPushed case to reducer

opr committed:
Ensure fullShippingAddressPushed is set when initialising cart store

opr committed:
Add fullShippingAddressPushed selector and default state entry

opr committed:
Add shippingAddressHasValidationErrors util function

opr committed:
Do not overwrite addresses when selecting a rate

opr committed:
Set whether full address has been pushed when saving address changes

opr committed:
In Shipping Methods Block, show correct component based on block setting

opr committed:
In Shipping Methods Block, show correct component based on block setting

opr committed:
Don't show from price if rates should be hidden until address entered

opr committed:
Check city validation errors to assert if shipping address is valid

opr committed:
Rename merchant.js to merchant.ts

opr committed:
Move local pickup functions to common merchant util

opr committed:
Update local pickup tests to use common merchant utils

opr committed:
Add test to ensure setting toggles in both blocks

opr committed:
Add navigating to settings and saving in merchant util

opr committed:
Create addPickupLocation merchant util

opr committed:
Add test for local pickup and require full address

opr committed:
Ensure setting displays correctly if cart is default but not checkout

opr committed:
Add tests to ensure core shipping settings update correctly

opr committed:
Show correct block in editor

opr committed:
Remove broken test from PR

opr committed:
Make sure correct conditions are met to show shipping options

opr committed:
Clean up updateAttributeInSiblingBlock

opr committed:
Ensure checkbox is checked during local pickup tests

opr committed:
Unset the checkbox when tests are finished running

wavvves reviewed (approved):
Tested out ok! `remove_shipping_settings()` might be a bit too big and confusing with all the iteration going on, but the comments make it easier.

opr committed:
Update checkout block fixture

opr committed:
Prevent error in unit tests

opr committed:
Import validation store key from constants

Required because importing from the index causes the validation data store to register twice

opr committed:
Update checkout terms test to wait for button not to be disabled

ralucaStan commented:
Mentioning that a part of this PR was merged [https://github.com/woocommerce/woocommerce-blocks/commit/510eec49f3f841bf61c531bf4460efe9a384f3c3#diff-a942a5198733d5fa79[…]34694286a6e49e4R420-R427](https://github.com/woocommerce/woocommerce-blocks/commit/510eec49f3f841bf61c531bf4460efe9a384f3c3#diff-a942a5198733d5fa79775b811f7cfe45bbf4d144bdf209a3334694286a6e49e4R420-R427) in the WC Blocks 9.8.0 release branch via this PR https://github.com/woocommerce/woocommerce-blocks/pull/8718. See the internal discussion p1678712946221229-slack-C8X6Q7XQU.

If this PR gets reverted at some point, the bug fix mentioned above should be brought to trunk.
